### PR TITLE
Graph meeting notifications nodejs sample

### PR DESCRIPTION
**Pull request by Prithvi-MSFT**

### CODE REVIEW

File: samples/graph-meeting-notification/nodejs/README.md [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+- nodejs
+extensions:
+ contentType: samples
+ createdDate: "24-09-2022 00:16:45"
+urlFragment: officedev-microsoft-teams-samples-graph-meeting-notification-nodejs
+---
+
+# Online meeting subscription
+
+This is a sample application which demonstrates use of online meeting subscription that will post notifications when user joined/left and when meeting start/end.
```
Comment (ChetanSharma-msft) [inline samples/graph-meeting-notification/nodejs/README.md:18]: Create a .gif image and add it here in Readme and update the sample.json image path.

File: samples/graph-meeting-notification/nodejs/README.md [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+## Concepts introduced in this sample
+- After sucessfully installation of bot in meeting you will get a welcome card and the subscription will be created for meeting it is installed in.
+
+![Welcome Card](Images/WelcomeCard.png)
+
+- After that when the metting gets started or user joins the meeting, notifications will be sent in chat:
+
+![Meeting Started](Images/MeetingStartedCard.png)
```
Comment (ChetanSharma-msft) [inline samples/graph-meeting-notification/nodejs/README.md:27]: Take a Teams full width screenshots for all the images and update at all the places. 
Also, .gif has the same full Teams width screenshots to make all the samples images consistent.

File: samples/graph-meeting-notification/nodejs/README.md [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+     npm start
+   ``` 
+
+ - Run ngrok - point to port 3978
+
+   ```bash
+     ngrok http -host-header=rewrite 3978
+
+## Instruction for manifest
+1. Fill any GUID for <APP-ID>. You can also put your MicrosoftAppId here.
```
Comment (ChetanSharma-msft) [inline samples/graph-meeting-notification/nodejs/README.md:118]: Fill any GUID for <GUID> in your manifest. As an alternative, you can also put your Microsoft App Id.

File: samples/graph-meeting-notification/nodejs/assets/sample.json [hunk 4]
```diff
@@ -0,0 +1,68 @@
+[
+  {
+    "name": "officedev-microsoft-teams-samples-graph-meeting-notification-nodejs",
+    "source": "officeDev",
+    "title": "Online meeting subscription",
```
Comment (ChetanSharma-msft) [inline samples/graph-meeting-notification/nodejs/assets/sample.json:5]: Online Teams Meeting Subscription

File: samples/graph-meeting-notification/nodejs/bots/meeting-notification-bot.js [hunk 5]
```diff
@@ -0,0 +1,144 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (ChetanSharma-msft) [inline samples/graph-meeting-notification/nodejs/bots/meeting-notification-bot.js:1]: Add header which includes file name as well.

Also, provide one line space before const.

File: samples/graph-meeting-notification/nodejs/bots/meeting-notification-bot.js [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+            this.addConversationReference(context.activity, decodedMeetingJoinUrl);
+            console.log('Running dialog with Message Activity.');
+            
+            await GraphHelper.createSubscription(meetingInfo.details.joinUrl);
+            await next();
+        });
+    }
+
+
```
Comment (ChetanSharma-msft) [inline samples/graph-meeting-notification/nodejs/bots/meeting-notification-bot.js:24]: Remove extra empty space and add method summary for all the methods.

File: samples/graph-meeting-notification/nodejs/helper/certHelper.js [hunk 7]
```diff
@@ -0,0 +1,130 @@
+// Copyright (c) Microsoft Corporation.
```
Comment (ChetanSharma-msft) [inline samples/graph-meeting-notification/nodejs/helper/certHelper.js:1]: We should rename this file as certificate-helper.

Also, update header which includes file name.

Remove extra empty line before const.

File: samples/graph-meeting-notification/nodejs/helper/certHelper.js [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+   * @param  {string} password - The password to use to protect the private key
+   */
+  createSelfSignedCertificateIfNotExists: async (
+    certPath,
+    keyPath,
+    password
+  ) => {
+    const certFullPath = path.join(__dirname, certPath);
+    return new Promise((resolve, reject) => {
+      if (!fs.existsSync(certFullPath)) {
```
Comment (ChetanSharma-msft) [inline samples/graph-meeting-notification/nodejs/helper/certHelper.js:44]: This code alignment does not looks correct.

File: samples/graph-meeting-notification/nodejs/helper/decryption-helper.js [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+class DecryptionHelper {
+    static async getDecryptedContent(notification) {
+
+        if (notification) {
+            
+            var payLoad = this.processEncryptedNotification(notification);
+            return payLoad;
+            
+        }
```
Comment (ChetanSharma-msft) [inline samples/graph-meeting-notification/nodejs/helper/decryption-helper.js:12]: We can remove extra empty lines inside if condition.

File: samples/graph-meeting-notification/nodejs/helper/graph-helper.js [hunk 10]
```diff
@@ -0,0 +1,97 @@
+
```
Comment (ChetanSharma-msft) [inline samples/graph-meeting-notification/nodejs/helper/graph-helper.js:1]: Remove first empty line
Update header as well

File: samples/graph-meeting-notification/nodejs/index.js [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+// Import required pckages
+const path = require('path');
+// Read botFilePath and botFileSecret from .env file.
+const ENV_FILE = path.join(__dirname, '.env');
+require('dotenv').config({ path: ENV_FILE });
+const restify = require('restify');
+const express = require('express');
+const { MeetingNotficationBot } = require('./bots/meeting-notification-bot');
+//const { MeetingNotficationBot } = require('./bots/meeting-notification-bot');
+const { DecryptionHelper } = require("./helper/decryption-helper");
```
Comment (ChetanSharma-msft) [inline samples/graph-meeting-notification/nodejs/index.js:12]: These line of code looks messy, kindly provide proper line spacing after file header and remove unwanted commented code.

File: samples/graph-meeting-notification/nodejs/index.js [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+    );
+
+    // Send a message to the user
+    await context.sendActivity('The bot encountered an error or bug.');
+    await context.sendActivity('To continue to run this bot, please fix the bot source code.');
+    // Clear out state
+    await conversationState.delete(context);
+};
+
+
```
Comment (ChetanSharma-msft) [inline samples/graph-meeting-notification/nodejs/index.js:57]: We can remove extra empty lines.

File: samples/graph-meeting-notification/nodejs/index.js [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+const memoryStorage = new MemoryStorage();
+
+// Create conversation and user state with in-memory storage provider.
+const conversationState = new ConversationState(memoryStorage);
+
+// Create the main dialog.
+//const dialog = new MainDialog();
+// Create the main dialog.
+const conversationReferences = {};
+// Create the bot that will handle incoming messages.
```
Comment (ChetanSharma-msft) [inline samples/graph-meeting-notification/nodejs/index.js:71]: One line space before it.

File: samples/graph-meeting-notification/nodejs/index.js [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+                                    title: "Member name",
+                                    value: member.Identity.User.DisplayName
+                                 }
+                               membersJoined.push(member);
+                            })
+
+                            await MeetingNotficationBot.DisplayMeetingUpdate(turnContext, "Members joined",membersJoined);
+                        }
+                        if(responsePayload['activeParticipants@remove'].length > 0) {
+                            var membersLeft = new Array();
```
Comment (ChetanSharma-msft) [inline samples/graph-meeting-notification/nodejs/index.js:122]: This code spacing is not correct, provide empty line before every if condition.

File: samples/graph-meeting-notification/nodejs/package.json [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+    "dependencies": {
+        "@microsoft/microsoft-graph-client": "^3.0.2",
+        "adaptivecards-templating": "^2.2.0",
+        "axios": "^0.21.1",
+        "botbuilder": "^4.16.0",
+        "crypto": "^1.0.1",
+        "dotenv": "^10.0.0",
+        "express": "^4.17.1",
+        "isomorphic-fetch": "^3.0.0",
+        "jszip": "^3.6.0",
```
Comment (ChetanSharma-msft) [inline samples/graph-meeting-notification/nodejs/package.json:27]: Are we using all of these packages, remove if not in use like   "jszip": "^3.6.0" etc.

**Review by ChetanSharma-msft — CHANGES_REQUESTED** [review]
Here are my review comments.

File: README.md [hunk 16]
```diff
@@ -362,6 +363,8 @@ The [Teams Toolkit](https://marketplace.visualstudio.com/items?itemName=TeamsDev
 [graph-activity-feed#js]:samples/graph-activity-feed/nodejs
 [graph-change-notification#cs]:samples/graph-change-notification/csharp
 [graph-change-notification#js]:samples/graph-change-notification/nodejs
+[graph-meeting-notification#cs]:samples/graph-meeting-notification/csharp
```
Comment (ChetanSharma-msft) [inline README.md:366]: Add these samples in pipeline as well.

**Review by ChetanSharma-msft — CHANGES_REQUESTED** [review]
Need pipeline file changes as well.

image

File: README.md (hunk 16 above)
Comment (ChetanSharma-msft) [reply README.md:366]: image

File: samples/graph-meeting-notification/nodejs/Manifest/manifest.json [hunk 17]
```diff
@@ -0,0 +1,51 @@
+{
+  "$schema": "https://developer.microsoft.com/en-us/json-schemas/teams/v1.11/MicrosoftTeams.schema.json",
+  "manifestVersion": "1.11",
```
[the remark this answers is not in the document]
Comment (Prithvi-MSFT) [reply samples/graph-meeting-notification/nodejs/Manifest/manifest.json:3]: These permissions do not work with v1.12/13 so used v1.11

**Review by ChetanSharma-msft — APPROVED** [review]
Changes looks fine, approving the PR.

File: samples/graph-meeting-notification/nodejs/helper/certificate-helper.js [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+const crypto = require('crypto');
+const pem = require('pem');
+/**
+ * Configures path to openssl.exe if needed
+ */
+function ensureOpenSsl() {
+  if (os.platform() === 'win32') {
+    const pathOpenSSL = process.env.OPENSSL_CONF
+      ? process.env.OPENSSL_CONF.replace('.cfg', '.exe')
+      : 'C:/Program Files/OpenSSL-Win64/bin/openssl.exe';
```
Comment (Wajeed-msft) [inline samples/graph-meeting-notification/nodejs/helper/certificate-helper.js:17]: Can we get this parameter as part of .env file?

**Review by Wajeed-msft — APPROVED** [review]

File: samples/graph-meeting-notification/nodejs/helper/certificate-helper.js (hunk 18 above)
Comment (ChetanSharma-msft) [reply samples/graph-meeting-notification/nodejs/helper/certificate-helper.js]: Added in .env file.

**Pull request merged**